src/devices/cpu/powerpc/ppcdrc.cpp

```diff
@@ -2261,7 +2261,20 @@ bool ppc_device::generate_opcode(drcuml_block &block, compiler_state *compiler,
 				UML_ADD(block, I0, mem(&m_core->tempaddr), (int16_t)G_SIMM(op) + 4 * (regnum - G_RD(op)));
 																							// add     i0,[tempaddr],simm + 4*(regnum-rd)
 				UML_CALLH(block, *m_read32align[m_core->mode]);         // callh   read32align
-				UML_MOV(block, R32(regnum), I0);                                        // mov     regnum,i0
+
+				if (regnum != G_RA(op) || regnum == 0 || ((m_cap & PPCCAP_4XX) && regnum == 31))
+				{
+					UML_MOV(block, R32(regnum), I0);                                        // mov     regnum,i0
+				}
+				else if (!(m_cap & PPCCAP_4XX) && m_flavor != PPC_MODEL_601)
+				{
+					// RA being in the range is an invalid form
+					// 403 and 405 manual allows writes where regnum != RA || regnum == 31, regnum == RA is skipped
+					// 601 user manual lists no special rules for R31 being valid but regnum == RA is skipped
+					// 603 and above just says it's an invalid form
+					// tropchnc (403GA) is known to use this behavior
+					UML_EXH(block, *m_exception[EXCEPTION_PROGRAM], 0x80000);          // exh    exception_program,0x80000
+				}
```

Comment:
The forms that generate an exception shouldn’t get into the loop at all.  The check should be done before the loop.

I’m pretty sure the chips that generate an exception will do so when RT is zero – from the architecture manual, “If RA is in the range of registers to be loaded, including the case in which RA=0, the instruction form is invalid.”  Even though RA=0 is treated as constant zero for D-form instructions, it’s still invalid.

I’d also guess that the 403/405/601 skip loading R0 when RA=0.

Reply:
Attempt 2. Doing it out of the loop makes a lot more sense.